File: src/minmpt.rs
use std::ptr::null_mut;
use thiserror::Error;

#[allow(dead_code)]
#[allow(non_camel_case_types)]
mod binding {
    include!(concat!(env!("OUT_DIR"), "/minmpt_bindings.rs"));
}

#[derive(Error, Debug)]
pub enum MinMPTError {
    #[error("Exceeded model context size")]
    ContextLimit,
    #[error("Invalid input")]
    InvalidInput,
    #[error("Internal failure")]
    Failure,
    #[error("Unknown error")]
    Unknown,
}

impl MinMPTError {
    fn from_code(code: i32) -> Self {
        match code as u32 {
            binding::MINMPT_FAILURE => Self::Failure,
            binding::MINMPT_INVALID => Self::InvalidInput,
            binding::MINMPT_CTX_LIMIT => Self::ContextLimit,
            _ => Self::Unknown,
        }
    }
}

#[derive(Default, Debug)]
pub struct MinMPTOptions {
    n_ctx_override: Option<usize>,
    n_threads: Option<u32>,
}

impl MinMPTOptions {
    pub fn override_n_ctx(self, n_ctx: usize) -> Self {
        Self {
            n_ctx_override: Some(n_ctx),
            ..self
        }
    }
    pub fn n_threads(self, n_threads: u32) -> Self {
        Self {
            n_threads: Some(n_threads),
            ..self
        }
    }
}

pub struct MinMPT {
    handle: binding::minmpt_handle,
}

#[allow(dead_code)]
impl MinMPT {
    pub fn load_model(
        path: &str,
        load_options: Option<MinMPTOptions>,
    ) -> Result<MinMPT, MinMPTError> {
        let mut me = MinMPT { handle: null_mut() };
        let load_options = load_options.unwrap_or_default();
        let href: *mut binding::minmpt_handle = &mut me.handle;
        let err = unsafe {
            binding::minmpt_load(
                href,
                path.as_bytes().as_ptr().cast(),
                path.as_bytes().len(),
                load_options.n_ctx_override.unwrap_or(0),
            )
        };
        if err == binding::MINMPT_OK as i32 {
            if let Some(nth) = load_options.n_threads {
                me.set_n_threads(nth);
            }
            Ok(me)
        } else {
            Err(MinMPTError::from_code(err))
        }
    }
    /// Creates a second model instance with its own kvcache but sharing the original model's
    /// weight data
    pub fn fork(&self) -> MinMPT {
        let mut child = MinMPT { handle: null_mut() };
        let selfref: binding::minmpt_handle = self.handle;
        let childref: *mut binding::minmpt_handle = &mut child.handle;
        unsafe{
            binding::minmpt_fork(selfref, childref);
        }
        child
    }
    pub fn set_n_threads(&self, n_threads: u32) {
        unsafe { binding::minmpt_set_n_threads(self.handle, n_threads) }
    }
    pub fn n_vocab(&self) -> usize {
        unsafe { binding::minmpt_n_vocab(self.handle) }
    }
    pub fn n_past(&self) -> usize {
        unsafe { binding::minmpt_n_past(self.handle) }
    }
    pub fn rewind(&self, n: usize) {
        unsafe { binding::minmpt_rewind(self.handle, n) }
    }
    pub fn n_ctx(&self) -> usize {
        unsafe { binding::minmpt_n_ctx(self.handle) }
    }
    pub fn reset_ctx(&mut self) {
        unsafe {
            binding::minmpt_reset_ctx(self.handle);
        }
    }
    pub fn eval(&mut self, ids: &[u32], logits_out: &mut Vec<f32>) -> Result<(), MinMPTError> {
        if ids.is_empty() {
            return Err(MinMPTError::InvalidInput)
        }
        logits_out.resize(self.n_vocab(), 0.0);
        let err = unsafe {
            binding::minmpt_eval_logits(
                self.handle,
                ids.as_ptr(),
                ids.len(),
                logits_out.as_mut_ptr(),
            )
        };
        if err == binding::MINMPT_OK as i32 {
            Ok(())
        } else {
            Err(MinMPTError::from_code(err))
        }
    }
}

impl Drop for MinMPT {
    fn drop(&mut self) {
        if !self.handle.is_null() {
            unsafe {
                binding::minmpt_free(self.handle);
            }
        }
    }
}

File: minmpt.cpp/minmpt.cpp
#include "minmpt.h"
#include "mpt.h"

#include <memory>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

struct minmpt_session {
  std::shared_ptr<mpt_model> model;
  std::unique_ptr<mpt_kvcache> kvcache;
  size_t mem_per_token = 0;
  int n_threads = 4;
  size_t n_past = 0;
};

static minmpt_session *from_handle(minmpt_handle h) {
  return reinterpret_cast<minmpt_session *>(h);
}

static const minmpt_session *from_handle_const(const minmpt_handle h) {
  return reinterpret_cast<const minmpt_session *>(h);
}

minmpt_error minmpt_load(minmpt_handle *handle, const char *filename,
                         size_t fnlen, size_t n_ctx_override) {
  auto modelp = new minmpt_session;
  std::string fn(filename, fnlen);
  try {
    modelp->model = std::make_shared<mpt_model>();
    if (mpt_model_load(fn, *modelp->model, n_ctx_override)) {
      modelp->kvcache = std::make_unique<mpt_kvcache>(*modelp->model);
      *handle = reinterpret_cast<minmpt_handle>(modelp);
      return MINMPT_OK;
    }
  } catch (...) {
    delete modelp;
  }
  return MINMPT_FAILURE;
}

void minmpt_fork(minmpt_handle handle, minmpt_handle *child) {
  auto modelp = from_handle_const(handle);
  auto newp = new minmpt_session;
  newp->mem_per_token = modelp->mem_per_token;
  newp->model = modelp->model;
  newp->kvcache = std::make_unique<mpt_kvcache>(*modelp->model);
  memcpy(newp->kvcache->memory_k->data, modelp->kvcache->memory_k->data,
         ggml_nbytes(newp->kvcache->memory_k));
  memcpy(newp->kvcache->memory_v->data, modelp->kvcache->memory_v->data,
         ggml_nbytes(newp->kvcache->memory_v));
  newp->n_past = modelp->n_past;
  *child = reinterpret_cast<minmpt_handle>(newp);
}

size_t minmpt_n_vocab(minmpt_handle handle) {
  auto modelp = from_handle(handle);
  return modelp->model->hparams.n_vocab;
  ;
}

size_t minmpt_n_ctx(minmpt_handle handle) {
  auto modelp = from_handle(handle);
  return modelp->model->hparams.n_ctx;
}

size_t minmpt_n_past(minmpt_handle handle) {
  auto modelp = from_handle(handle);
  return modelp->n_past;
}

void minmpt_rewind(minmpt_handle handle, size_t n) {
  auto modelp = from_handle(handle);
  if (modelp->n_past > n) {
    modelp->n_past -= n;
  } else {
    modelp->n_past = 0;
  }
}

void minmpt_reset_ctx(minmpt_handle handle) {
  auto modelp = from_handle(handle);
  modelp->n_past = 0;
}

void minmpt_set_n_threads(minmpt_handle handle, unsigned int n_threads) {
  auto modelp = from_handle(handle);
  modelp->n_threads = n_threads > 0 ? n_threads : 4;
}

minmpt_error minmpt_eval_logits(minmpt_handle handle, const uint32_t *tokens,
                                size_t n_tokens, float *logits) {
  auto modelp = from_handle(handle);
  if (modelp->n_past + n_tokens > (size_t)modelp->model->hparams.n_ctx) {
    return MINMPT_CTX_LIMIT;
  }
  if (modelp->mem_per_token == 0) {
    std::vector<float> dummy_logits;
    mpt_eval_cpp(*modelp->model, *modelp->kvcache, modelp->n_threads, 0,
                 {1, 2, 3, 4}, dummy_logits, modelp->mem_per_token);
  }
  if (!mpt_eval(*modelp->model, *modelp->kvcache, modelp->n_threads,
                modelp->n_past, tokens, n_tokens, logits,
                modelp->mem_per_token)) {
    printf("Failed to predict\n");
    return MINMPT_FAILURE;
  }
  modelp->n_past += n_tokens;
  return MINMPT_OK;
}
void minmpt_free(minmpt_handle handle) {
  auto modelp = from_handle(handle);
  delete modelp;
}
